### DISCUSSION TITLE
Alternative Hardware Idea

### DISCUSSION THREAD

**Discussion by Mike-K8LH** [TomNisbet/TommyPROM#64] [Ideas]
Greetings, Tom.  Would you consider modifying the hardware by replacing the serial-to-parallel 74HC164 ICs with a pair of cascaded 8-bit parallel 74HC574 ICs?  The change would free up a few pins on the Nano and setting the address lines should be much faster.  One 'downside' is that breadboard or PCB routing would be a little bit more involved.

Setting the address lines is accomplished by writing two bytes to the '574s.  Note that the **/OE** pin is maintained high and the **/CE** pin should be high when setting the address latches.  Set the data lines with the high address byte, strobe the **/OE** pin low then high, set data lines with low address byte, strobe the **/OE** pin low then high.  After that, you can take the **/OE** pin low when required in order to read a byte from the ROM, however if you take it from low to high you'll corrupt the address latches.   

Cheerful regards, Mike, K8LH (Michigan, USA)
![Flash Programmer #3A](https://github.com/TomNisbet/TommyPROM/assets/38607118/11bc97f7-43e8-499a-bfb1-5ffa11e8406e)
![Flash Programmer #3B](https://github.com/TomNisbet/TommyPROM/assets/38607118/0f9b02d4-4cb9-4a0d-a580-eedd77f0145c)

**Comment by TomNisbet** [comment]
Hi Mike,

You could gain a bit of speed by using flip-flops instead of shift registers, but you would only free up one Arduino pin from the current design.  As you point out, the wiring would be more complicated because the 8 data line would now need to route to the EEPROM and both address flip-flops.  The software change would be trivial - the PromAddressDriver::setAddressRegister function would load the registers from the data bus instead of clocking in the values a bit at a time.

The current design uses three Arduino pins to control the address registers.  Each register has a clock pin to load data and there is a common data pin that feeds them both.  With flip-flops for the address registers, the clock pins would still be needed, but the data pin would be free.

I'm not sure I understand your proposed drawing for the registers.  Are you connecting the Aurduino OE pin to the EEPROM and to the registers?  How would different values be loaded into each register?

I like the idea of onboard MT3806s to provide programming voltages. I just connect a bench supply when programming non-standard chips.

Have you looked at the [TommyPROM32 board](https://tomnisbet.github.io/TommyPROM/docs/pcb#tommyprom32)?  I built that to program the SST39SF chips, but it has a modular jumper configuration to accommodate other chips as well.  It is similar to what you are doing with the chip jumpers, but more flexible.  I need to put up a picture of the new spin of the board that matches the schematic.  The new board has two programming voltages instead of one, and adds a few more power connection pins.

If you build a version with flip-flops, it will be interesting to see the actual speed difference when programming a chip.  [This trace](https://tomnisbet.github.io/TommyPROM/docs/28C256-notes) shows that the existing code takes about 70us to set both address lines.  That doesn't add up to too much for most chips because sequential writes means that only the lower address is set.  But the SST39SF chips write a command sequence for each byte, so the address writes really start to add up.

**Comment by Mike-K8LH** [comment]
Hi Tom,

The '574 based solution actually frees up 3 pins by eliminating the CLK-HI, CLK-LO, and the AD-BIT pins.  The /OE line is doing double-duty by driving the ROM and the clock input on the cascaded '574 latches.  The data lines are only connected to the first '574 and the ROM.  The outputs from the first '574 connect to the second '574 inputs and the ROM low address lines.  The outputs from the second '574 connect to the ROM high address lines.  There is a bit more routing involved compared to a '164 based solution but it's not difficult.

![nvRAM chip](https://github.com/TomNisbet/TommyPROM/assets/38607118/03978a1f-f668-465a-9101-1fb0500a291a)

Loading an address into the cascaded '574 latch ICs is accomplished by clocking the high address byte into the cascade followed by the low address byte.  Bytes are loaded into and shifted along the cascade during each clock strobe.  I don't have a scope but I estimate the example putAddr() function below takes a little more than 3us.

```
 /***************************************************************************
  *  write address to the cascaded 74HC574 latch ICs (hi address first)     *
  *                                                                         */
  void putAddr()                              //
  { busOut();                                 //
    PORTB = (PORTB & 0xFC) | (addr.hi >> 6);  // set D7..D6 bits
    PORTD = (PORTD & 0x03) | (addr.hi << 2);  // set D5..D0 bits
    oe(0); oe(1);                             // strobe latches
    PORTB = (PORTB & 0xFC) | (addr.lo >> 6);  // set D7..D6 bits
    PORTD = (PORTD & 0x03) | (addr.lo << 2);  // set D5..D0 bits
    oe(0); oe(1);                             // strobe latches
  }                                           // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Another example uses less memory at the cost of a couple extra instruction cycles;
```
 /***************************************************************************
  *  write address to the cascaded 74HC574 latch ICs (hi address first)     *
  *                                                                         */
  void putAddr()                              // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  { putData(addr.hi); oe(0); oe(1);           // load 'hi' address byte
    putData(addr.lo); oe(0); oe(1);           // load 'lo' address byre
  }                                           // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 /***************************************************************************
  *  write 'data' byte onto the data bus                                    *
  *                                                                         */
  void putData(byte data)                     // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  { busOut();                                 //
    PORTB = (PORTB & 0xFC) | (data >> 6);     // set D7..D6 pins
    PORTD = (PORTD & 0x03) | (data << 2);     // set D5..D0 pins
  }                                           // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Interestingly, I can write bytes to Flash ROM's (4 writes per byte) between incoming serial characters at 115,200 baud (~86us intervals).

I have looked at the TommyPROM32 design.  It's very nice.  

Cheerful regards, Mike

**Comment by TomNisbet** [comment]
I didn't realize that the output of one register feeds the input of the other.  It makes much more sense now.  I had to draw out a timing diagram to convince myself that your design would actually work!  The two things that weren't immediately obvious:

- CS allows the address to be set without bus contention from the EEPROM 
- The address register doesn't get corrupted until the *end* of the read cycle, when OE returns HIGH

That's a clever design.

You could probably use the TommyPROM code by replacing PromAddressDriver::setAddress with your code.  There is no need for any of the additional logic in that file because you always set both registers.

You will also need to check for places where the PROM is being read to make sure the address is stable.  For example, PromDevice28C::waitForWriteCycleEnd rads the chip twice at the same address without setting the address register either time.  This would need to be modified to maintain the address register.

Also, there may be cases where the address is being set while the PROM CS is enabled.

Another option would be to use a different pin for the address register load instead of OE.  This still saves two pins over my shift register hardware.  Are you using those other pins?  The TommyPROM32 design can address an SST39SF040 chip and still has D13 as a spare pin.

**Reply by Mike-K8LH** [reply]
When I came across TommyPROM a few years ago I tried to modify the setAddress driver for my Flash Programmer shield hardware but "no joy".  I think I had trouble with the C++ structure.   

I have four free pins on my programmer.  I really only need to program 5v EEPROMs and 5v Flash ROMs but I thought perhaps you might be able to take advantage of four free pins.

Regards...

![BB ZIF-32 Adapter Demo](https://github.com/TomNisbet/TommyPROM/assets/38607118/ee8c9b91-d3b0-413c-8c1b-9408b9161224)

**Comment by TomNisbet** [comment]
Also, I looked up the U63764 chip in your photo - never heard of that one before.  Reminds me of the hybrid SSD and spinning drives that were popular for a while.  I'll have to see if I can dig one of those up to add to my list of supported chips.

**Reply by Mike-K8LH** [reply]
That's a really nice 8K nvRAM chip (standard JEDEC RAM pinout).  A friend of mine gave me a handful of them (new old stock) and I'd love to recommend them to people but I just have not found them anywhere.  I'm using them in a little Apple-1 emulator (below).

![Crab_Apple_1a](https://github.com/TomNisbet/TommyPROM/assets/38607118/e8e0afae-431d-4ed9-9c93-4e7d81d26893)

**Reply by TomNisbet** [reply]
I've taken it as a bit of a challenge to see how many different chip families I can program.  It's interesting to research the different chips and make them work.  I'll have to scour Ebay for some of these nvRAM chips.  In addition to the ZMD chips, there is another manufacturer who makes the STK12C68 and STK14C88.  Also the FM16W08 and FM18W08 ferroelectric series.

I suspect that your board didn't work with my code due to the multiple read issue I mentioned above.  I've you ever want to give it another try, I'm happy to help debug.
